wallet-new/src/Cardano/Wallet/Kernel/Transactions.hs

```diff
@@ -0,0 +1,279 @@
+        shouldRetry _ _                                     = return True
+
+-- | Creates a new 'TxAux' without submitting it to the network.
+newTransaction :: ActiveWallet
+               -> IO Address
+               -- ^ A computation to generate change 'Address'es.
+               -> (Address -> Either CoinSelHardErr SafeSigner)
+               -- ^ A function to sign each Address
+               -> CoinSelectionOptions
+               -- ^ The options describing how to tune the coin selection.
+               -> HdAccountId
+               -- ^ The source @(root, account) from where the payment was
+               -- originated
+               -> NonEmpty (Address, Coin)
+               -- ^ The payees
+               -> IO (Either NewTransactionError TxAux)
+newTransaction ActiveWallet{..} genChangeAddr signAddress options accountId payees = do
+    snapshot <- getWalletSnapshot walletPassive
+    let toTxOuts = fmap (\(a,c) -> TxOutAux (TxOut a c))
+    -- We generate a new change address outside the 'MonadRandom' monad in which
+    -- coin selection runs, so that we keep our \"external\" model consistent
+    -- by generating it once, before calling coin selection. Specifically, we
+    -- pass this potential change address from the outside.
+    -- One disadvantage of this approach is that in case of failure of the
+    -- coin-selection we might be left with such dangling address, but in the
+    -- future we could easily add a function to delete an address, keeping the
+    -- possible number of orphans at bay.
+    let mkTx = mkStdTx walletProtocolMagic signAddress
+    -- | FIXME(adn) This number was computed out of the work Matt Noonan did
+    -- on the size estimation. See [CBR-318].
+    let maxInputs = 350
+    let availableUtxo = accountAvailableUtxo snapshot accountId
+    res <- flip runReaderT payees . buildPayment $
+           CoinSelection.random options
+                                (liftIO genChangeAddr)
+                                mkTx
+                                maxInputs
+                                (toTxOuts payees)
+                                availableUtxo
+    case res of
+         Left err -> return . Left . CoinSelectionFailed $ err
+         Right t  -> return . Right $ t
+
+-- | Special monad used to process the payments, which randomness is derived
+-- from a fixed seed obtained from hashing the payees. This guarantees that
+-- when we estimate the fees and later create a transaction, the coin selection
+-- will always yield the same value, making the process externally-predicatable.
+
+newtype PayMonad a = PayMonad {
+      buildPayment :: ReaderT (NonEmpty (Address, Coin)) IO a
+    } deriving (Functor, Applicative, Monad, MonadIO, MonadReader (NonEmpty (Address, Coin)))
+
+-- | \"Invalid\" 'MonadRandom' instance for 'PayMonad' which generates
+-- randomness using the hash of 'NonEmpty (Address, Coin)' as fixed seed.
```

Comment:
Er, not sure I understand this. Is this ti make coin selection deterministic? Are we running coin selection in this monad? What do we use this for?